Improve OpenMP thread scaling

paulromano opened the pull request:
This pull request makes two important changes that can greatly improve scalability on multi- and many-core architectures when OpenMP is used.

1. Currently, nested particle geometry coordinates are dynamically allocated as a particle moves through geometry through the Particle % coord linked list. The problem with this is that heap allocation (`allocate(...)` in Fortran, `malloc` in C) can cause thread contention because system runtime libraries often have to use a lock to synchronize access. I've now changed Particle % coord to be statically allocated with a maximum size of 5.
2. Writes to global tallies during transport are currently synchronized using !$omp atomic. A potential problem with this is that after one thread updates a global tally, when another thread goes to write to it, it causes a cache miss due to invalidation (assuming a coherent cache). Thus, as you add more threads, you can end up with more cache misses and therefore lower performance. The "standard" way to handle this would be to use `!$omp parallel do reduction(....)`. However, as both @davidozog and I found, the reduction clause doesn't actually protect writes to the specified variables, meaning you still need an atomic or critical (I consider this a bug of both gfortran and ifort). In this PR, I created threadprivate copies of each global tally and reduce them myself manually, thus avoiding the cache misses due to invalidation.

As one anecdote of how important these effects are, on the Xeon Phi 7120a, these changes lead to a 4x speedup when using 244 threads on the Hoogenboom-Martin benchmark.

wbinventor commented:
Do you believe the 5 is a large enough number to limit the statically
allocated coords linked list? Have you tried this with the BEAVRS
benchmark? If I recall, @nhorelik's construction of that benchmark resulted
in more than 5 nested universe levels, but I may be wrong.
On Jul 2, 2015 8:55 PM, "Paul Romano" <notifications@github.com> wrote:

> This pull request makes two important changes that can greatly improve
> scalability on multi- and many-core architectures when OpenMP is used.
>
>    1. Currently, nested particle geometry coordinates are dynamically
>    allocated as a particle moves through geometry through the Particle % coord
>    linked list. The problem with this is that heap allocation (
>    allocate(...) in Fortran, malloc in C) can cause thread contention
>    because system runtime libraries often have to use a lock to synchronize
>    access. I've now changed Particle % coord to be statically allocated with a
>    maximum size of 5.
>    2. Writes to global tallies during transport are currently
>    synchronized using !$omp atomic. A potential problem with this is that
>    after one thread updates a global tally, when another thread goes to write
>    to it, it causes a cache miss due to invalidation (assuming a coherent
>    cache). Thus, as you add more threads, you can end up with more cache
>    misses and therefore lower performance. The "standard" way to handle this
>    would be to use !$omp parallel do reduction(....). However, as both
>    @davidozog <https://github.com/davidozog> and I found, the reduction
>    clause doesn't actually protect writes to the specified variables, meaning
>    you still need an atomic or critical (I consider this a bug of both
>    gfortran and ifort). In this PR, I created threadprivate copies of each
>    global tally and reduce them myself manually, thus avoiding the cache
>    misses due to invalidation.
>
> As one anecdote of how important these effects are, on the Xeon Phi 7120a,
> these changes lead to a 4x speedup when using 244 threads on the
> Hoogenboom-Martin benchmark.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mit-crpg/openmc/pull/408
> Commit Summary
>
>    - Use omp reduction instead of atomic on global tallies
>    - Static allocation for Particle % coord instead of linked list. Not
>    all tests
>    - Fix bug in print_particle
>    - Fix handling of reflective boundary conditions
>    - Update results for test_filter_universe
>    - Call LocalCoord % reset() in gfortran-4.6 compatible manner
>    - Rather than use reduction clause in !$omp do (which doesn't work in
>    gfortran or
>    - Only accumulate k-effective global tallies if eigenvalue calculation
>    - Rename threadprivate global tally variables and add a few comments
>
> File Changes
>
>    - *M* src/constants.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-0> (9)
>    - *M* src/eigenvalue.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-1> (26)
>    - *M* src/geometry.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-2> (292)
>    - *M* src/global.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-3> (17)
>    - *M* src/output.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-4> (70)
>    - *M* src/particle_header.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-5> (61)
>    - *M* src/particle_restart.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-6> (8)
>    - *M* src/physics.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-7> (28)
>    - *M* src/plot.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-8> (104)
>    - *M* src/source.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-9> (12)
>    - *M* src/tally.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-10> (60)
>    - *M* src/track_output.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-11> (2)
>    - *M* src/tracking.F90
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-12> (53)
>    - *M* tests/test_filter_universe/results_true.dat
>    <https://github.com/mit-crpg/openmc/pull/408/files#diff-13> (12)
>
> Patch Links:
>
>    - https://github.com/mit-crpg/openmc/pull/408.patch
>    - https://github.com/mit-crpg/openmc/pull/408.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/408>.
>

smharper commented:
I don't think 5 is enough, either.  I used more than 5 when I modelled the FHTR, and I probably used more than 5 when I modelled the Czech LR-0.

I'm also opposed to the idea of having a max limit at all.  It feels like we're going back in time to the days when memory couldn't be dynamically allocated.  Maybe some user will find a scenario where they need 100 coordinate levels to easily model their system or get optimal performance out of OpenMC.

What if we used allocatable lists that keep a little buffer space?  We could have a system where if you have to allocate space for a coordinate level, then you allocate space for five.  Would we retain the performance boost that way?

paulromano commented:
Yes, it seems BEAVRS does need more than 5. @cjosey just pointed out to me that he had to increase it to 8.

@smharper To sympathize, I am also opposed to have max limits in general. However, in this case it seems to be performance critical to avoid heap allocations. As you suggest, we could use dynamically growing arrays that get allocated in chunks, but this comes at the expense of implementation complexity. And it still leaves potential allocations on the heap, particularly at the most nested levels of the geometry which would be allocated/deallocated frequently. For now, I think I'd prefer to stick with a statically allocated coord array with a reasonably high MAX_COORD.

wbinventor commented:
I can certainly buy the static allocation argument for a
performance-critical data structure like local coords. I do think that
increasing the default value to 8-10 is more appropriate. What about also
making it possible to set this as a compile time macro? Such that something
like the following is possible:

```bash
cmake . -DMAX_COORDS=25
```

This would be fine for C/C++ and I assume the same is true for Fortran.

Also, I like the idea of adding a check to the ``Geometry`` class in the
Python API for overly nested universe hierachies. Perhaps when a
``GeometryFile`` exports to XML, it could ask the ``Geometry`` to traverse
the CSG tree to determine the maximum depth? And then compare this to the
default value, and inform the user if it is greater to rebuild OpenMC with
the compile time macro.
On Jul 4, 2015 6:36 PM, "Paul Romano" <notifications@github.com> wrote:

> Yes, it seems BEAVRS does need more than 5. @cjosey
> <https://github.com/cjosey> just pointed out to me that he had to
> increase it to 8.
>
> @smharper <https://github.com/smharper> To sympathize, I am also opposed
> to have max limits in general. However, in this case it seems to be
> performance critical to avoid heap allocations. As you suggest, we could
> use dynamically growing arrays that get allocated in chunks, but this comes
> at the expense of implementation complexity. And it still leaves potential
> allocations on the heap, particularly at the most nested levels of the
> geometry which would be allocated/deallocated frequently. For now, I think
> I'd prefer to stick with a statically allocated coord array with a
> reasonably high MAX_COORD.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/408#issuecomment-118569350>.
>

smharper commented:
@paulromano, fair enough.  4x speedup is quite significant, after-all.

I like @wbinventor's idea of setting a max coords at compile time.  I also agree that we should traverse the tree and alert the user if they need more coordinates, but I think it should be done at the OpenMC level.

On another note, why did the "test_filter_universe" results change?

paulromano commented:
I've changed it so that there is now a -Dmaxcoord= CMake option, which defaults to 10.  I also added a maximum_levels function in geometry.F90  which loops through the geometry to determine the maximum number of nested coordinate levels. A check is made using this function now.

paulromano commented on src/geometry.F90:
```diff
@@ -1333,24 +1314,19 @@ subroutine distance_to_boundary(p, dist, surface_crossed, lattice_translation)
     end do LEVEL_LOOP
 
-    ! Move particle to appropriate coordinate level
-    if (associated(final_coord)) p % coord => final_coord
```
Removing this line is essentially why test_filter_universe changed. Since the universe filter looked for `p % coord % universe` before, if distance_to_boundary caused the coordinate level to change (even though the particle hasn't really "left" that level yet), it would result in a different universe being scored to. Now, since the check is on `p % coord(p % n_coord) % universe`, the apparent universe won't change until the next call to find_cell, which doesn't happen until after tracklength tallies are scored... I know this is a little confusing, but I do think the behavior now is more correct, or at least consistent.

cjosey commented:
I've developed an extension to this PR on my private repo.  Instead of statically allocating based on a compile flag, I instead allocate once based on @paulromano 's checking code.  Unfortunately, his checking code gets the depth wrong right now, (for the older version of  BEAVRS I carry around to run all my testing, it calculates 4 instead of 8), but if I force it to be 8, I get nearly identical performance with none of the original caveats (static allocation).  Once I pop this onto a public repo instead of a private one and figure out github's UI, I'll make you all aware of it.

Below is a graph indicating why I have strong support for this PR.  This is BEAVRS run without tallies or Shannon entropy, on a dual-processor, 36 core in total node, using Intel ifort 15.0.1.  BEAVRS geometry is a bit more complex than H-M, so the differences are a lot more extreme.  I saw a factor of 50 speedup on the Xeon Phi.
![graph](https://cloud.githubusercontent.com/assets/354602/8535882/12231452-2418-11e5-86d9-268d9cf8a1f9.png)

paulromano commented:
@cjosey Thanks for demonstrating the importance of the changes. I'll look into the depth calculation to see what's wrong.

wbinventor commented on docs/source/usersguide/install.rst:
```diff
@@ -160,6 +160,13 @@ openmp
 
+coverage
+  Compile and link code instrumented for coverage analysis. This is typically
+  used in conjunction with gcov.
```
perhaps include a link to "gcov"?

wbinventor commented on src/geometry.F90:
```diff
@@ -130,6 +125,7 @@ recursive subroutine find_cell(p, found, search_cells)
     logical,        intent(inout) :: found
     integer,        optional      :: search_cells(:)
     integer :: i                    ! index over cells
+    integer :: j
```
perhaps ``n_coord`` or ``coord_num`` would be a more transparent var name for the new "j" local vars in a lot of these routines?

wbinventor commented on src/geometry.F90:
```diff
@@ -76,20 +76,18 @@ subroutine check_cell_overlap(p)
 
     integer :: i                       ! cell loop index on a level
+    integer :: j
+    integer :: n_coord
```
Oops, I see that you already have a ``n_coord`` here, so perhaps it won't be as easy to intuitively name "j". If there isn't a good name, at least add a comment explaining what this var is for here

wbinventor commented on src/geometry.F90:
```diff
@@ -189,27 +187,28 @@ recursive subroutine find_cell(p, found, search_cells)
         ! CELL CONTAINS LOWER UNIVERSE, RECURSIVELY FIND CELL
 
         ! Create new level of coordinates
```
This comment should now be something like "Index lower level ..." instead of "Create new level ..."

wbinventor commented on src/geometry.F90:
```diff
@@ -220,40 +219,41 @@ recursive subroutine find_cell(p, found, search_cells)
-        i_xyz = lat % get_indices(p % coord % xyz + TINY_BIT * p % coord % uvw)
+        i_xyz = lat % get_indices(p % coord(j) % xyz + TINY_BIT * p % coord(j) % uvw)
 
         ! Create new level of coordinates
```
ditto reword this comment

wbinventor commented on src/geometry.F90:
```diff
@@ -1907,5 +1885,82 @@ recursive subroutine count_instance(univ)
+      if (c % type == CELL_FILL) then
+
+        next_univ => universes(c % fill)
+        levels_below = maximum_levels(next_univ)
```
Should this not be:

```fortran
levels_below = max(levels_below, maximum_levels(next_univ))
```

as it is below for ``CELL_LATTICE`` types? Otherwise I think this will fail if it finds a ``CELL_FILL`` with fewer levels *after* inspecting a ``CELL_LATTICE`` with more levels.

paulromano replied:
Ah ha, that's the bug! Thanks @wbinventor!

wbinventor replied:
Haha, no problem, that's what I thought. I think this was just a sign that I've spent far too much time on CSG stuffs these past few years.

wbinventor commented on src/output.F90:
```diff
@@ -273,39 +272,34 @@ subroutine print_particle(p)
+    do i = 1, p % n_coord
       ! Print level
-      write(ou,*) '  Level ' // trim(to_str(i))
+      write(ou,*) '  Level ' // trim(to_str(i - 1))
```
Will this fail (segfault) if ``i == ``?

wbinventor commented on src/particle_header.F90:
```diff
@@ -42,8 +41,8 @@ module particle_header
     ! Particle coordinates
-    type(LocalCoord), pointer :: coord0 => null() ! coordinates on universe 0
-    type(LocalCoord), pointer :: coord  => null() ! coordinates on lowest universe
+    integer :: n_coord                   ! number of current coordinates
```
Can these be var declarations have matching indentations to aid readability?

paulromano commented:
Any impediment to getting this merged? I'm not aware of any outstanding issues.

wbinventor commented:
I have no objections but didn't want to pull the trigger before @cjosey
and/or @smharper confirm all is well.
On Jul 13, 2015 5:07 AM, "Paul Romano" <notifications@github.com> wrote:

> Any impediment to getting this merged? I'm not aware of any outstanding
> issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mit-crpg/openmc/pull/408#issuecomment-120904559>.
>

cjosey commented:
I've been extensively testing this PR, and have found no further bugs.  I tried implementing a version that did single allocation to eliminate the drawbacks, as I mentioned earlier, but it only works on Intel compilers.  It appears that gfortran follows standards when it comes to allocatables inside of derived types used as threadprivate objects and ifort does not.
